awx/main/scheduler/task_manager_instances.py

```diff
@@ -0,0 +1,236 @@
+# Copyright (c) 2015 Ansible, Inc.
+# All Rights Reserved
+
+# Python
+import logging
+import heapq
+from types import SimpleNamespace
+
+# Django
+from django.conf import settings
+
+# AWX
+from awx.main.scheduler.dependency_graph import DependencyGraph
+from awx.main.managers import InstanceGroupManager
+from awx.main.models import (
+    Instance,
+    InstanceGroup,
+)
+
+
+logger = logging.getLogger('awx.main.scheduler')
+
+
+class PrioritizableNode:
+    def __init__(self, remaining_capacity, hostname):
+        self.remaining_capacity = remaining_capacity
+        self.hostname = hostname
+
+    def __gt__(self, other):
+        return self.remaining_capacity > other.remaining_capacity
+
+    def __lt__(self, other):
+        return self.remaining_capacity < other.remaining_capacity
+
+    def __eq__(self, other):
+        if hasattr(other, 'hostname'):
+            other = other.hostname
+        return self.hostname == other
+
+    def __hash__(self):
+        return hash(self.hostname)
+
+    def __str__(self):
+        return f"{self.hostname}: {self.remaining_capacity} remaining capacity"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class PrioritizedNodes:
+    def __init__(self):
+        self.nodes = set()
+
+    def add(self, node):
+        self.nodes.add(node)
+
+    def __getitem__(self, hostname):
+        if hostname not in self.nodes:
+            raise KeyError(f"No control node with hostname {hostname} found")
+        for node in self.nodes:
+            if node == hostname:
+                return node
+
+    def update(self, node):
+        """The __hash__ function for PrioritizableNodes just compares hostnames.
+
+        We can add/remove the node to update object in PrioritizedNodes to have right capacity.
+        """
+        self.nodes.remove(node)
+        self.nodes.add(node)
+
+    def best_node(self):
+        heap = list(self.nodes)
+        heapq.heapify(heap)
+        return heapq.nlargest(1, heap).pop()
+
+    def __str__(self):
+        return f"{[n for n in self.nodes]}"
+
+
+class TaskManagerInstances:
+    def __init__(self):
+        realinstances = Instance.objects.filter(hostname__isnull=False, enabled=True).exclude(node_type='hop').prefetch_related('rampart_groups')
+        self.instances_partial = dict()
+        self.control_nodes = PrioritizedNodes()
+
+        for instance in realinstances:
+            self.instances_partial[instance.hostname] = SimpleNamespace(
+                obj=instance,
+                node_type=instance.node_type,
+                remaining_capacity=instance.remaining_capacity,
```

Comment:
There are going to be :dragon: s here

https://github.com/ansible/awx/blob/2de5ffc8d94d592498c32b8c3d68a82470d57763/awx/main/models/ha.py#L146-L152

You're using a property on the Instance model which is _very_ inefficient. It queries the list of jobs anew, which is really really counter to the purpose of the task manager. The task manager has complex logic to load instances, instance groups, and jobs. It is _most_ important that we don't re-load jobs multiple times. You can't just create a new data structure for instance capacity because the hard part is integrating it with the list of jobs that the task manager fetches.